src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,254 @@
+        public ProvenBlockHeaderRepository(Network network, string folder, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            Directory.CreateDirectory(folder);
+
+            this.dbreeze = new DBreezeEngine(folder);
+            this.network = network;
+        }
+
+        /// <inheritdoc />
+        public Task InitializeAsync()
+        {
+            Task task = Task.Run(() =>
+            {
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    if (this.TipHashHeight != null) return;
+
+                    var hashHeight = new HashHeightPair(this.network.GetGenesis().GetHash(), 0);
+
+                    this.SetTip(transaction, hashHeight);
+
+                    transaction.Commit();
+
+                    this.TipHashHeight = hashHeight;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<List<ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Task<List<ProvenBlockHeader>> task = Task.Run(() =>
+            {
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    transaction.SynchronizeTables(ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.logger.LogTrace("Loading ProvenBlockHeaders from block height '{0}' to '{1}  from the database.",
+                        fromBlockHeight, toBlockHeight);
+
+                    var headers = new List<ProvenBlockHeader>();
+
+                    for (int i = fromBlockHeight; i <= toBlockHeight; i++)
+                    {
+                        Row<byte[], ProvenBlockHeader> row =
+                            transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, i.ToBytes(false));
+
+                        if (row.Exists)
+                        {
+                            headers.Add(row.Value);
+                        }
+                        else
+                        {
+                            this.logger.LogTrace("ProvenBlockHeader block height ({0}) does not exist in the database.", i);
+                            headers.Add(null);
+                        }
+                    }
+
+                    return headers;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            IEnumerable<ProvenBlockHeader> headers =  await this.GetAsync(blockHeight, blockHeight).ConfigureAwait(false);
+
+            return headers.First();
```

Comment:
first or default so we return null in case item doesn't exits instead of throwing

Reply:
not fixed

Reply:
Changed to FirstOrDefault().